### ISSUE TITLE
Build successful but methods are not exported

### ISSUE THREAD

**Issue by bbartels** [3F/DllExport#59]
### How to reproduce
Open Solution in VS and build. 

Build is successful and build log indicates that the Exported methods are assembled correctly, though when using dumpbin they do not show up.

Code repository: https://github.com/bbartels/K8055Simulator/tree/DLLExportIssues
    
### What version and selected configuration ?

DLLExport.1.5.2

### Can you provide detailed log of your build ?

https://pastebin.com/gUBpwPVv

I am in the middle of porting my project from R.Giesecke's UnmanagedExports to DLLExport.1.5.2, didn't have an issue with UnmanagedExports. (Except not being able to find ILdasm at times, which is the reason why I am porting)

**Comment by 3F** [comment]
Did you try latest v1.6-RC ?

29.11.2017 3:28, Benjamin Bartels пишет:

**Comment by Genteure** [comment]
From the log, it seems some methods got successful exported.
Which reminds me of about a month ago, I had a similar problem.

I have a project that has two source file contains methods need to export.
One of them successful export every time I build, the other one doesn't get exported every time.
I checked build result using dnSpy, the DllExport Attribute didn't get removed.

I was able to find a kind simple solution: just [change the order of `Compile` `ItemGroup` in the `csproj` file](https://github.com/Genteure/com.genteure.cqp.whatanime/blob/master/com.genteure.cqp.whatanime/com.genteure.cqp.whatanime.csproj#L62) so the compiler compiles source files in a different order. which somehow magically solved the problem.

Not sure if is the same problem, but you can give a try.

**Comment by bbartels** [comment]
@3F I have tried the 1.6 Beta but haven't tried the RC yet. Will do so as soon as I can.

**Comment by Genteure** [comment]
@3F You can check out my repository, I pushed 3 new branches with build log and result binary.

## [dllexport-16b2](https://github.com/Genteure/com.genteure.cqp.whatanime/tree/dllexport-16b2)

Current working example, everything works fine.

## [dllexport-16b2-bug](https://github.com/Genteure/com.genteure.cqp.whatanime/tree/dllexport-16b2-bug)

Default build order, methods in `Main.cs` doesn't get exported.

## [dllexport-16rc](https://github.com/Genteure/com.genteure.cqp.whatanime/tree/dllexport-16rc)

Default build order with DllExport 1.6.0-RC, methods in `Main.cs` doesn't get exported.

![DllExport Configure](https://user-images.githubusercontent.com/11240579/33389515-0f264c6a-d56e-11e7-94d1-56ca393bb2de.png)
![Build Result](https://user-images.githubusercontent.com/11240579/33389380-9f9fa62a-d56d-11e7-87dc-f7c64e1e7ff3.png)

**Comment by 3F** [comment]
@Genteure Thanks for the detailed info ! I will look this as soon as possible. 

By the way, what your system locale ? it seems you have incorrect layout :( a little bit, but...

**Comment by Genteure** [comment]
My system locale is zh_CN Simplified Chinese, on Windows 10.
I changed my Visual Studio's language to English so msbuild log is in English.

**Comment by 3F** [comment]
@Genteure Good, I'll try your locale in VM :)

**Comment by 3F** [comment]
@Genteure 

well, yes 1.6-RC indicates 2 errors when PE-checking is used (your branch [dllexport-16rc](https://github.com/Genteure/com.genteure.cqp.whatanime/tree/dllexport-16rc)):

```
RunIlAsm: Something went wrong. We can't find '2' exports from PE32/PE32+ module '\com.genteure.cqp.whatanime\bin\Debug\com.genteure.cqp.whatanime.dll': _eventEnable ; _eventGroupMsg	com.genteure.cqp.whatanime	\packages\DllExport.1.6.0-RC\tools\net.r_eg.DllExport.targets		
RunIlAsm: The number (2) of exports from PE32/PE32+ module '\com.genteure.cqp.whatanime\bin\Debug\com.genteure.cqp.whatanime.dll' is not equal to number (4) from IL code.	com.genteure.cqp.whatanime	\packages\DllExport.1.6.0-RC\tools\net.r_eg.DllExport.targets		
```

And I've found the following problem from your code. The problem mainly because of another decompiled view for `extern` methods (where DllImport attr) from your NativeMethods static class:

```csharp
  .class abstract auto ansi sealed nested private beforefieldinit 'NativeMethods'
         extends ['mscorlib']'System'.'Object'
  {
    ...
    .method public hidebysig static pinvokeimpl("CQP.DLL" winapi) 
            string  'CQ_getRecord'(int32 'AuthCode',
                                   string 'File',
                                   string 'Format') cil managed preservesig
    {
    }
    ...
```

In general, we have incorrect nesting level of subclasses because of our current logic via stack-based levels.

Well, need to think...

While I'm thinking, you can also try https://github.com/3F/Conari that generates binding at runtime, and moreover...

**Comment by 3F** [comment]
@bbartels I've also checked your https://github.com/bbartels/K8055Simulator/commit/2cbbe82b626468d5637f17901358c22c919555da and **for v1.6-RC** I have no problems:

x64:

```
00000001	0003EB9E	0000	0004022A	OpenDevice
00000002	0003EBAA	0001	0004021E	CloseDevice
00000003	0003EBB6	0002	00040276	ReadAnalogChannel
00000004	0003EBC2	0003	00040259	ReadAllAnalog
00000005	0003EBCE	0004	000401F7	ClearAnalogChannel
00000006	0003EBDA	0005	000401D8	ClearAllAnalog
00000007	0003EBE6	0006	00040245	OutputAnalogChannel
00000008	0003EBF2	0007	00040235	OutputAllAnalog
00000009	0003EBFE	0008	00040300	SetAnalogChannel
0000000A	0003EC0A	0009	000402E5	SetAllAnalog
0000000B	0003EC16	000A	000401E7	ClearAllDigital
0000000C	0003EC22	000B	0004020A	ClearDigitalChannel
0000000D	0003EC2E	000C	000402F2	SetAllDigital
0000000E	0003EC3A	000D	0004033E	SetDigitalChannel
0000000F	0003EC46	000E	00040367	WriteAllDigital
00000010	0003EC52	000F	000402AC	ReadDigitalChannel
00000011	0003EC5E	0010	00040267	ReadAllDigital
00000012	0003EC6A	0011	000402A0	ReadCounter
00000013	0003EC76	0012	000402D8	ResetCounter
00000014	0003EC82	0013	00040327	SetCounterDebounceTime
00000015	0003EC8E	0014	00040350	SetDigitalInputChannel
00000016	0003EC9A	0015	000402BF	ReadDigitalOutputChannel
00000017	0003ECA6	0016	00040311	SetAnalogInputChannel
00000018	0003ECB2	0017	00040288	ReadAnalogOutputChannel
```

**Comment by 3F** [comment]
Ok, I've plan to view this problem later, after planned v1.6 (should be released for this month).

For today, anyone else may also try to use https://github.com/3F/Conari to avoid similar @Genteure's problem.

**Comment by 3F** [comment]
Fixed and included in planned v1.6.1  

Use CI-Build-92 for tests.
